_episodes/08-defensive.md

```diff
@@ -237,40 +342,250 @@ $ git push origin master
 >
 {: .callout}
 
-> ## Add your own assertions
+
+## Logging
+
+So far we've considered how to make our programs halt or handle the situation when things go wrong.
+Another option in our defensive programming toolkit is to have our programs report their own activity.
+
+For example,
+let's say we're working with relative humidity data.
+We wouldn't typically expect to encounter any values over 100%, but it is physically possible.
+Rather than halt the program if a value over 100% occurs,
+we might therefore want our program to simply report the maximum relative humidity.
+We can then decide whether to trust the output or not
+(e.g. a value of 100.1% might be ok but not 150%).
+
+To do this reporting,
+our first instinct might be to add a `print` statement to the program.
+
+~~~
+rh_data = np.array([1.5, 20.4, 100.1, 76.3, 54.4])
+rh_max = rh_data.max()
+print(f'The maximum relative humidity was {rh_max}%')
+~~~
+{: .language-python}
+
+~~~
+The maximum relative humidity was 100.1%
+~~~
+{: .output}
+
+The problem with this approach is that information printed to the screen is lost
+once we close our command line session.
+Constantly adding, removing or commenting out `print` statements is also tedious and error-prone.
+
+A better approach is to use a logging framework,
+such as Python’s `logging` library.
+This lets us leave debugging statements in our code and turn them on or off at will.
+Let's start by replacing our `print` statement with a `logging` command.
+
+~~~
+import logging
+
+rh_data = np.array([1.5, 20.4, 100.1, 76.3, 54.4])
+rh_max = rh_data.max()
+logging.debug(f'The maximum relative humidity was {rh_max}%')
+~~~
+{: .language-python}
+
+Whoops!
+There's no output because by default the logging library only reports information
+at the "warning" level and above.
+In order of increasing severity, the available levels are:
+
+- `debug`: very detailed information used for localizing errors.
+- `info`: confirmation that things are working as expected.
+- `warning`: something unexpected happened, but the program will keep going.
+- `error`: something has gone badly wrong, but the program hasn’t hurt anything.
+- `critcal`: potential loss of data, security breach, etc.
+
+If we want to see the output from less severe levels (i.e. turn our debugging statements on),
+we'd need to change the minimum level in the logging configuration.
+We can also provide the name of a file to write the logging information to,
+so that it isn't lost when we finish our command line session.
+
+~~~
+# for loop only required in notebooks
+for handler in logging.root.handlers[:]:
+    logging.root.removeHandler(handler)
+    
+logging.basicConfig(level=logging.debug, filename='log.txt')) 
+
+rh_data = np.array([1.5, 20.4, 100.1, 76.3, 54.4])
+rh_max = rh_data.max()
+logging.debug(f'The maximum relative humidity was {rh_max}%')
+~~~
+{: .language-python}
+
+(The for loop is needed to turn off the background logging the notebook does itself.
```

Comment:
Suggested addition:
By setting the logging level to "debug", our output "log.txt" file will now capture all logging information with a flag of 'debug' or higher - that is, all logging outputs will be written to our log file.